File: copy_insert.hs
-- Copy-insert function
copyInsert :: a -> [[a]] -> [[[a]]]
copyInsert _ [] = [[[]]]
copyInsert a (x:xs) = map (insertIntoSublist a) (copyInsert a xs) ++ [insertIntoFirstSublist a x : xs]

-- Helper to insert element into each sublist
insertIntoSublist :: a -> [[a]] -> [[a]]
insertIntoSublist a l = [a] : l

-- Helper to insert element into the first sublist
insertIntoFirstSublist :: a -> [a] -> [a]
insertIntoFirstSublist a list = a : list

-- Main function to test the utility functions
main :: IO ()
main = do
    let a = 4
    let l = [[1], [2, 3]]
    print (copyInsert a l)

File: copy_insert.py
def copy_insert(a, l):
    """
    Given `l`, a list of lists, insert `a` in each sublist of `l`,
    and return the list of all modifications of `l`.
    """
    if not l:
        return [[a]]
    
    first_list = l[0]
    rest_lists = l[1:]

    # Recursive call to handle the rest of the lists
    rest_modifications = copy_insert(a, rest_lists)

    # Generate modifications where `a` is inserted into each sublist of `rest_modifications`
    modified_with_first = [ [a] + first_list ] + rest_lists
    rest_results = [ [first_list] + result for result in rest_modifications ]

    return [modified_with_first] + rest_results

# Example usage
a = 4
l = [[1], [2, 3]]
print(copy_insert(a, l))
